Fix dock dragging and dropping

matthewwithanm opened the pull request:
This fixes #16769, which described an issue whereby trying to drop into
a closed dock would cause the dock to repeatedly open and close.

I tried to describe in comments why this was happening and how we can
make sure to avoid it.

The solution adds another DOM measurement which I'm not too thrilled
about but I profiled and it didn't seem to be an issue.

matthewwithanm committed:
Fix dock dragging and dropping

This fixes #16769, which described an issue whereby trying to drop into
a closed dock would cause the dock to repeatedly open and close.

I tried to describe in comments why this was happening and how we can
make sure to avoid it.

The solution adds another DOM measurement which I'm not too thrilled
about but I profiled and it didn't seem to be an issue.

as-cii reviewed (approved):
Looks good @matthewwithanm! Measuring one more time shouldn't be a huge concern because we were already calling `getBoundingClientRect` inside `pointWithinHoverArea`, so unless I am missing something this pull request won't add extra reflows.

Thanks for jumping on this, please merge at will once tests are green again. ⚡️

daviwil reviewed (approved):
Thanks a lot Matthew!

matthewwithanm committed:
Make sure there's a single source of truth for dock hover state

Previously, the workspace's idea of the hovered dock and the docks'
themselves could be out of sync.
